src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/InvalidPiersCheck.java

```diff
@@ -0,0 +1,362 @@
+                configValue -> HighwayTag.valueOf(configValue.toUpperCase()));
+    }
+
+    /**
+     * This function will validate if the supplied atlas object is valid for the check. Valid object
+     * for the check is a master edge with man_made=pier tag and does not have an area=yes tag.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return {@code true} if this object should be checked
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge && ((Edge) object).isMasterEdge() && ManMadeTag.isPier(object)
+                && HAS_NO_AREA_TAG.test(object) && !this.isFlagged(object.getOsmIdentifier());
+    }
+
+    /**
+     * Overriding this method to not skip piers.
+     *
+     * @return true since we need the check to accept piers
+     */
+    @Override
+    protected boolean acceptPier()
+    {
+        return true;
+    }
+
+    /**
+     * This is the actual function that will check to see whether the object needs to be flagged.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @SuppressWarnings("squid:S3655")
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+        // We will mark the edge as flagged as we will only be looking at all way sectioned edges of
+        // a way once
+        this.markAsFlagged(edge.getOsmIdentifier());
+        // Collect all master edges that form the OSM way
+        final Set<Edge> edgesFormingOSMWay = new OsmWayWalker(edge).collectEdges().stream()
+                .filter(Edge::isMasterEdge).collect(Collectors.toSet());
+        final List<Edge> listOfEdgesFormingOSMWay = new ArrayList<>(edgesFormingOSMWay);
+        // We need to sort the edges to build the geometry of the OSM way in a deterministic manner
+        listOfEdgesFormingOSMWay.sort(Comparator.comparingLong(AtlasObject::getIdentifier));
+        // Build a polygon using the locations in the edges
+        final Set<Location> locationsInOsmWay = new HashSet<>();
+        listOfEdgesFormingOSMWay
+                .forEach(osmWayEdge -> locationsInOsmWay.addAll(osmWayEdge.asPolyLine()));
+        final Polygon osmWayAsPolygon = new Polygon(locationsInOsmWay);
+        // Check if the OSM way has linear geometry or polygonal geometry
+        final boolean isPolygonal = this.hasPolygonalGeometry(listOfEdgesFormingOSMWay, edge);
+        final int instructionIndex;
+        // We can flag the edge if it has a highway tag with the right priority or is a polygonal
+        // pier with building tag or is a polygonal pier with amenity=ferry_terminal
+        if ((HighwayTag.highwayTag(edge).isPresent() && HighwayTag.highwayTag(edge).get()
```

Comment:
Are there cases where a flag can contain multiple issues? Ideally, with each issue, we could concatenate the instructions and flag them all at the end

Reply:
Since checking for overlappingHighway and intersecting/overlapping buildings is computationally heavy tasks, I short circuit the check and return the flag if  the pier has a highway tag or is polygon with building tag/amenity=ferry terminal tag. If we want to include all the conditions in the instructions, then we need to check for intersections/overlaps even if it has a highway tag. Thoughts?

Reply:
Ahh ic, I'm cool with that. We can revisit it later !